compiler/rustc_macros/src/session_diagnostic.rs

```diff
@@ -363,9 +355,9 @@ struct SessionDiagnosticDeriveBuilder<'a> {
     /// Slug is a mandatory part of the struct attribute as corresponds to the Fluent message that
     /// has the actual diagnostic message.
     slug: Option<(String, proc_macro::Span)>,
-    /// Error codes are a mandatory part of the struct attribute. Slugs may replace error codes
-    /// in future but it is desirable to mandate error codes until such a time.
-    code: Option<(String, proc_macro::Span)>,
+    /// Error codes are a optional part of the struct attribute - this is only set to detect
```

Comment:
thought: feels like codes should be mandatory for now until we switch over wholesale to slugs at some point in the future, since we still want to generate the error-index

Reply:
I didn't do it that way because so many of our current diagnostics don't have error codes, and adding them increases the friction for converting to the derive (having to reserve an error code number makes the a pull request more likely to have conflict and then you need to write the description for the error). 

Any error that currently has a code should be ported to a derive that has that code, but I don't personally mind if the derive doesn't force codes to be added where they didn't exist before, we can still make a mapping from error codes to slugs even if not all slugs had an error code.

Longer-term, it would be nice if we could have the error code index be generated from mandatory documentation comments on derive structs.

I don't feel too strongly here though, I can drop that commit (making error codes optional) if others from @rust-lang/wg-diagnostics prefer.

Reply:
Oh, I was under the impression codes were already mandatory. Never mind!!

Reply:
Codes were only mandatory when using the derive, but it was only being used for something like a dozen diagnostics.

Reply:
Yea, there is a fairly high number of calls to `span_err` without a code and we've not had a strong policy around them. This seems good to me and we can address it separately